src/Model/LazyBSONArray.php

```diff
@@ -0,0 +1,273 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace MongoDB\Model;
+
+use AppendIterator;
+use ArrayAccess;
+use ArrayIterator;
+use CallbackFilterIterator;
+use Iterator;
+use IteratorAggregate;
+use JsonSerializable;
+use MongoDB\BSON\PackedArray;
+use MongoDB\BSON\Serializable;
+use MongoDB\Codec\CodecLibrary;
+use MongoDB\Codec\LazyBSONArrayCodec;
+use MongoDB\Codec\LazyBSONDocumentCodec;
+use MongoDB\Exception\InvalidArgumentException;
+use ReturnTypeWillChange;
+
+use function array_key_exists;
+use function array_keys;
+use function array_merge;
+use function array_values;
+use function is_array;
+use function is_object;
+use function iterator_to_array;
+use function max;
+use function sprintf;
+use function trigger_error;
+
+use const E_USER_WARNING;
+
+/**
+ * Model class for a BSON array.
+ *
+ * The internal data will be filtered through array_values() during BSON
+ * serialization to ensure that it becomes a BSON array.
+ *
+ * @api
+ */
+class LazyBSONArray implements ArrayAccess, IteratorAggregate, JsonSerializable, Serializable
+{
+    /** @var ?CodecLibrary */
+    private $library;
+
+    /** @var PackedArray */
+    private $bson;
+
+    /** @var array<int, mixed> */
+    private $read = [];
+
+    /** @var array<int, true> */
+    private $notFound = [];
+
+    /** @var array<int, mixed> */
+    private $set = [];
+
+    /** @var array<int, true> */
+    private $unset = [];
+
+    /** @var bool */
+    private $entirePackedArrayRead = false;
+
+    /**
+     * Deep clone this lazy array.
+     */
+    public function __clone()
+    {
+        $this->bson = clone $this->bson;
+
+        foreach ($this->set as $key => $value) {
+            if (is_object($value)) {
+                $this->set[$key] = clone $value;
+            }
+        }
+    }
+
+    /**
+     * Constructs a lazy BSON array.
+     *
+     * @param PackedArray|array|null $input An input for a lazy array.
+     *        When given a BSON array, this is treated as input. For lists
+     *        this constructs a new BSON array using fromPHP.
+     */
+    public function __construct($input = null, ?CodecLibrary $library = null)
+    {
+        if ($input === null) {
+            $this->bson = PackedArray::fromPHP([]);
+        } elseif ($input instanceof PackedArray) {
+            $this->bson = $input;
+        } elseif (is_array($input)) {
+            $this->bson = PackedArray::fromPHP(array_values($input));
+        } else {
+            throw InvalidArgumentException::invalidType('input', $input, [PackedArray::class, 'array', 'null']);
+        }
+
+        $this->library = $library;
+    }
+
+    public function bsonSerialize(): array
+    {
+        // Always use LazyBSONArrayCodec for BSON serialisation
+        $codec = new LazyBSONArrayCodec();
+        $codec->attachLibrary($this->getLibrary());
+
+        // @psalm-suppress InvalidReturnStatement
```

Comment:
`toPHP` returns `array|object`, so we specifically have to ignore psalm errors here as we know that a `PackedArray::toPHP()` will always return an array in the absence of a type map.